File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionReduce.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Stream;

public class TransformerFunctionReduce extends TransformerFunction {
    public TransformerFunctionReduce() {
        super(FunctionDescription.of(
                Map.of(
                        "to", ArgumentType.of(ArgType.Transformer).position(0).defaultIsNull(true),
                        "identity", ArgumentType.of(ArgType.Any).position(1).defaultIsNull(true)
                )
        ));
    }
    @Override
    public CompletionStage<Object> apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null)
            return null;
        var identity = context.getJsonElement("identity");
        var to = context.getJsonElement("to", false);

        var i = new AtomicInteger(0);
        return ((Stream<Object>)streamer.stream())
                .reduce(identity, (acc, x) -> context.transformItem(to, x, i.getAndIncrement(), "##accumulator", acc));
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionReduce.ts
import { AsyncSequence, asyncSequenceOf } from "@wortise/sequency";
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { ArgType } from "./common/ArgType";

class TransformerFunctionReduce extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        to: { type: ArgType.Transformer, position: 0, defaultIsNull: true },
        identity: { type: ArgType.Any, position: 1, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null) return null;
    const identity = await context.getJsonElement("identity");
    const to = await context.getJsonElement("to", false);

    return (
      asyncSequenceOf(asyncSequenceOf(identity), streamer.stream()).flatten() as AsyncSequence<any>
    ).reduceIndexed(async (i, acc, x) => {
      return context.transformItem(to, x, i - 1, "##accumulator", acc);
    });
  }
}

export default TransformerFunctionReduce;
